File: 0368-largest-divisible-subset/0368-largest-divisible-subset.py
class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        n = len(nums)
        nums.sort()
        dp = [1] * n
        h = [i for i in range(n)]
        maxi = 0
        for i in range(1, n):
            for p in range(i):
                if not nums[i] % nums[p]:
                    if dp[p] + 1 > dp[i]:
                            dp[i] = dp[p] + 1
                            h[i] = p
                            if dp[i] > dp[maxi]:
                                maxi = i

        res = []
        i = maxi
        while h[i] != i:
            res.append(nums[i])
            i = h[i]
        res.append(nums[i])
        return res
            

        

File: 0368-largest-divisible-subset/0368-largest-divisible-subset.java
class Solution {
    public List<Integer> largestDivisibleSubset(int[] nums) {
        int n = nums.length;
        int [] dp = new int[n];
        Arrays.fill(dp, 1);
        Arrays.sort(nums);
        int []par = new int[n];
        for (int i = 0; i < n; i++) par[i] = i;

        int maxIdx = 0;
        int maxLen = 1;

        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[i] % nums[j] == 0) {
                    if (dp[i] < 1 + dp[j]) {
                        par[i] = j;
                        dp[i] = 1 + dp[j];
                        if (dp[i] > dp[maxIdx]) maxIdx = i;
                    }
                }
            }
        }

        List<Integer> ans = new ArrayList<>();

        while (maxIdx != par[maxIdx]) {
            ans.add(nums[maxIdx]);
            maxIdx = par[maxIdx];            
        }

        ans.add(nums[maxIdx]);

        return ans;

    }
}